include/boost/geometry/algorithms/azimuth.hpp

```diff
@@ -0,0 +1,226 @@
+// Boost.Geometry
+
+// Copyright (c) 2007-2012 Barend Gehrels, Amsterdam, the Netherlands.
+
+// This file was modified by Oracle on 2014-2021.
+// Modifications copyright (c) 2014-2021, Oracle and/or its affiliates.
+// Contributed and/or modified by Vissarion Fysikopoulos, on behalf of Oracle
+// Contributed and/or modified by Adam Wulkiewicz, on behalf of Oracle
+
+// Use, modification and distribution is subject to the Boost Software License,
+// Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_GEOMETRY_ALGORITHMS_AZIMUTH_HPP
+#define BOOST_GEOMETRY_ALGORITHMS_AZIMUTH_HPP
+
+
+#include <boost/geometry/algorithms/not_implemented.hpp>
+
+#include <boost/geometry/core/cs.hpp>
+#include <boost/geometry/core/access.hpp>
+#include <boost/geometry/core/radian_access.hpp>
+#include <boost/geometry/core/tag.hpp>
+#include <boost/geometry/core/tags.hpp>
+
+#include <boost/geometry/strategies/azimuth.hpp>
+#include <boost/geometry/strategies/default_strategy.hpp>
+#include <boost/geometry/strategies/azimuth/cartesian.hpp>
+#include <boost/geometry/strategies/azimuth/geographic.hpp>
+#include <boost/geometry/strategies/azimuth/spherical.hpp>
+
+#include <boost/geometry/util/math.hpp>
+
+
+namespace boost { namespace geometry
+{
+
+
+#ifndef DOXYGEN_NO_DETAIL
+namespace detail
+{
+       
+} // namespace detail
+#endif // DOXYGEN_NO_DETAIL
+
+
+#ifndef DOXYGEN_NO_DISPATCH
+namespace dispatch
+{
+
+template
+<
+    typename Geometry1, typename Geometry2,
+    typename Tag1 = typename tag<Geometry1>::type,
+    typename Tag2 = typename tag<Geometry2>::type
+>
+struct azimuth : not_implemented<Tag1, Tag2>
+{};
+
+template <typename Point1, typename Point2>
+struct azimuth<Point1, Point2, point_tag, point_tag>
+{
+    template <typename Strategy>
+    static auto apply(Point1 const& p1, Point2 const& p2, Strategy const& strategy)
+    {
+        typedef typename decltype(strategy.azimuth())::template result_type
```

Comment:
you can avoid it by using `auto` below, right?

Reply:
Not without changes in azimuth strategies. Azimuth strategy `CalculationType`, Point1's coordinate type and Point2's coordinate type can be different types and right now this strategy `apply()` requires all arguments of the same type.

It would be doable if azimuth strategy `apply()` allowed to pass arguments of arbitrary type, was doing the conversion to the calculation type inside and returning the result from the function instead of using output argument for that. Then there would be a problem with the overloads of `apply()` allowing to calculate both forward and reverse azimuths at the same time. These functions would have to have different name than `apply()` which would be ok I guess since there already is `apply_reverse()`.

So yes, it could be avoided but would require more changes. I think this is a good idea in general to remove all of these utilities like `result_type`, `state_type`, etc. and replace them with a mechanism using `auto`. But this would mean changes in the interfaces of strategies. If we agreed to do this I'd rather do this in a different PR.

Reply:
sure, better doing this in separate PR